CMakeLists.txt

```diff
@@ -1,110 +1,463 @@
+        find_package(Threads REQUIRED)
+        target_link_libraries(abc_interface
+            INTERFACE ${CMAKE_THREAD_LIBS_INIT}
+        )
+    endif()
+endif()
+
+if(ABC_USE_PIC)
+    message(STATUS "Compiling position independent code")
+    set(CMAKE_POSITION_INDEPENDENT_CODE ON)
+endif()
+
+if(ABC_USE_LIBSTDCXX OR NOT ABC_USE_NAMESPACE)
+    if(NOT CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")
+        message(STATUS "Using explicit -lstdc++")
+        target_link_libraries(abc_interface
+            INTERFACE stdc++
+        )
+    endif()
+endif()
+
+define_property(GLOBAL
+    PROPERTY LIBABC_SOURCES
+    BRIEF_DOCS "Sources of libabc"
+    FULL_DOCS "All source files of libabc"
+)
+
+define_property(GLOBAL
+    PROPERTY ABC_SOURCES
+    BRIEF_DOCS "Sources of executables of abc"
+    FULL_DOCS "All source files of abc"
+)
+
+function(abc_libabc_add_sources)
+    cmake_parse_arguments(AAM "" "NAME" "SOURCES" ${ARGN})
+    file(RELATIVE_PATH RELDIR "${abc_SOURCE_DIR}" "${CMAKE_CURRENT_LIST_DIR}")
+    set(RELSOURCES)
+    foreach(SOURCE ${AAM_SOURCES})
+        list(APPEND RELSOURCES "${RELDIR}/${SOURCE}")
     endforeach()
+    source_group("${RELDIR}" FILES "${AAM_SOURCES}")
+    set_property(GLOBAL APPEND PROPERTY LIBABC_SOURCES ${RELSOURCES})
 endfunction()
 
-project(abc)
+function(abc_add_executable)
+    cmake_parse_arguments(AAE "" "NAME" "SOURCES" ${ARGN})
+    file(RELATIVE_PATH RELDIR "${abc_SOURCE_DIR}" "${CMAKE_CURRENT_LIST_DIR}")
+    set(RELSOURCES)
+    foreach(SOURCE ${AAE_SOURCES})
+        list(APPEND RELSOURCES "${RELDIR}${SOURCE}")
+    endforeach()
+    if(ABC_USE_NAMESPACE)
+        set_source_files_properties(${AAE_SOURCES} PROPERTIES LANGUAGE CXX)
+    endif()
+    add_executable("${AAE_NAME}" ${AAE_SOURCES})
+    target_link_libraries("${AAE_NAME}" PUBLIC libabc)
+    set_property(GLOBAL APPEND PROPERTY ABC_SOURCES ${RELSOURCES})
+endfunction()
 
-if(READLINE_FOUND MATCHES TRUE)
-    addprefix(ABC_READLINE_INCLUDES_FLAGS "-I" ${READLINE_INCLUDE})
-    string(REPLACE ";" " " ABC_READLINE_INCLUDES_FLAGS "${ABC_READLINE_INCLUDES_FLAGS}")
-    list(APPEND ABC_READLINE_FLAGS "ABC_READLINE_INCLUDES=${ABC_READLINE_INCLUDES_FLAGS}")
+foreach(ABC_MODULE ${ABC_MODULES})
+    add_subdirectory("${ABC_MODULE}")
+endforeach()
 
-    string(REPLACE ";" " " ABC_READLINE_LIBRARIES_FLAGS "${READLINE_LIBRARIES}")
-    list(APPEND ABC_READLINE_FLAGS "ABC_READLINE_LIBRARIES=${ABC_READLINE_LIBRARIES_FLAGS}")
-elseif(READLINE_FOUND MATCHES FALSE)
-    list(APPEND ABC_READLINE_FLAGS "ABC_USE_NO_READLINE=1")
-endif()
+get_property(LIBABC_SOURCES GLOBAL PROPERTY LIBABC_SOURCES)
+get_property(ABC_SOURCES GLOBAL PROPERTY ABC_SOURCES)
 
 if(ABC_USE_NAMESPACE)
-    set(ABC_USE_NAMESPACE_FLAGS "ABC_USE_NAMESPACE=${ABC_USE_NAMESPACE}")
-endif()
-
-if( APPLE )
-    set(make_env ${CMAKE_COMMAND} -E env SDKROOT=${CMAKE_OSX_SYSROOT})
-endif()
-
-# run make to extract compiler options, linker options and list of source files
```

Comment:
I'm not sure I understand the requirement deriving this; can you elaborate a little bit here?

Reply:
Sorry, that ^^ question was in regards to cmake docs, where it says:

"...  It is intended to be set locally by the user creating a build tree. "

https://cmake.org/cmake/help/latest/variable/CMAKE_OSX_SYSROOT.html